src/maxflow.rs

```diff
@@ -1 +1,251 @@
+use std::iter;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
 
+// Maybe it should be in a separate module
+pub trait MfCapacity:
+    Copy + Ord + Add<Output = Self> + AddAssign + Sub<Output = Self> + SubAssign
+{
+    fn zero() -> Self;
+    fn max_value() -> Self;
+}
+
+impl MfCapacity for i32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i32::MAX
+    }
+}
+
+impl MfCapacity for i64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i64::MAX
+    }
+}
+
+impl MfCapacity for u32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u32::MAX
+    }
+}
+
+impl MfCapacity for u64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u64::MAX
+    }
+}
+
+impl<Cap> MfGraph<Cap>
+where
+    Cap: MfCapacity,
+{
+    pub fn new(n: usize) -> MfGraph<Cap> {
+        MfGraph {
+            _n: n,
+            pos: Vec::new(),
+            g: iter::repeat_with(Vec::new).take(n).collect(),
+        }
+    }
+
+    pub fn add_edge(&mut self, from: usize, to: usize, cap: Cap) -> usize {
+        assert!(from < self._n);
+        assert!(to < self._n);
+        assert!(Cap::zero() < cap);
+        let m = self.pos.len();
+        self.pos.push((from, self.g[from].len()));
+        let rev = self.g[to].len();
+        self.g[from].push(_Edge { to, rev, cap });
+        let rev = self.g[from].len() - 1;
+        self.g[to].push(_Edge {
+            to: from,
+            rev,
+            cap: Cap::zero(),
+        });
+        m
+    }
+}
+
+struct Edge<Cap> {
+    from: usize,
+    to: usize,
+    cap: Cap,
+    flow: Cap,
+}
+
+impl<Cap> MfGraph<Cap>
+where
+    Cap: MfCapacity,
+{
+    fn get_edge(&self, i: usize) -> Edge<Cap> {
+        let m = self.pos.len();
+        assert!(i < m);
+        let _e = &self.g[self.pos[i].0][self.pos[i].1];
+        let _re = &self.g[_e.to][_e.rev];
+        Edge {
+            from: self.pos[i].0,
+            to: _e.to,
+            cap: _e.cap + _re.cap,
+            flow: _re.cap,
+        }
+    }
+    fn edges(&self) -> Vec<Edge<Cap>> {
+        let m = self.pos.len();
+        (0..m).map(|i| self.get_edge(i)).collect()
+    }
+    fn change_edge(&mut self, i: usize, new_cap: Cap, new_flow: Cap) {
+        let m = self.pos.len();
+        assert!(i < m);
+        assert!(Cap::zero() < new_flow && new_flow <= new_cap);
+        let (to, rev) = {
+            let _e = &mut self.g[self.pos[i].0][self.pos[i].1];
+            _e.cap = new_cap - new_flow;
+            (_e.to, _e.rev)
+        };
+        let _re = &mut self.g[to][rev];
+        _re.cap = new_flow;
+    }
+
+    pub fn flow(&mut self, s: usize, t: usize) -> Cap {
+        self.flow_with_capacity(s, t, Cap::max_value())
+    }
+    pub fn flow_with_capacity(&mut self, s: usize, t: usize, flow_limit: Cap) -> Cap {
+        let n_ = self._n;
+        assert!(s < n_);
+        assert!(t < n_);
```

Comment:
```suggestion
        assert!(t < n_);
        assert_ne!(s, t);
```

Reply:
I think I should not add this assertion, because the original documentation does not say it disallows the same vertices, and by definition we can return a valid value `flow_limit`.

Reply:
By definition in the document appendix.html, we should return 0 I believe, which is very confusing as you might have noticed.

> g(v, f) = Sum{...} - Sum{...}

This implies the sum of g(v, f) over all vertices is zero.

> g(v, f) = g(v, f') holds for all vertices v other than s and t.

Since s = t, this also implies g(t, f) = g(t, f').

> g(t,f′)−g(t,f) is maximized under these conditions. It returns this g(t,f′)−g(t,f)

Since g(t, f) = g(t, f'), the returned value should be 0.

Reply:
I finally understood.  It seems to be true that it should return 0.  Doesn't the current code?

Reply:
I tried the original library with a simple example and it returned `MAX_VALUE` instead...

Reply:
Reported in https://github.com/atcoder/ac-library/issues/5

Reply:
ACL fixed it :)